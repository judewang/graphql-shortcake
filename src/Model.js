/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
