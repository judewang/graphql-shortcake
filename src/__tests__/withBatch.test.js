import _ from 'lodash';
import database from './database';
import Model from '..';

class Batch extends Model {
  static database = database;

  static tableName = 'batch';

  static hasOperator = true;

  static columns = {
    name: { type: String },
  }
}

class Logger extends Batch {
  static hasOperator = false;

  static hasTimestamps = false;
}

class Cache extends Batch {
  static softDelete = false;

  static hasOperator = false;

  static hasTimestamps = false;
}

describe('Batch', () => {
  beforeAll(async () => {
    await database.schema.dropTableIfExists('batch');

    await database.schema.createTable('batch', (table) => {
      table.increments();
      table.string('name');
      table.timestamps();
      table.integer('created_by');
      table.integer('updated_by');
      table.timestamp('deleted_at');
      table.integer('deleted_by');
    });
  });

  describe('insert', () => {
    const rows = _.map(_.range(1, 10000), idx => ({ name: `x${idx}` }));
    it('successfully insert', async () => {
      const results = await Cache.batchInsert(rows);
      expect(results.length).toBe(9999);
    });

    it('successfully insert with operator', async () => {
      const results = await Batch.batchInsert(rows, 99);
      expect(results.length).toBe(9999);
    });

    it('operator is required', async () => {
      await expect(Batch.batchInsert(rows))
        .rejects.toEqual(new Error('operator is required'));
    });
  });

  describe('update', () => {
    const rows = ['1', '3', '5'];
    it('successfully update', async () => {
      await Cache.batchUpdate(rows, { name: 'cupcake' });
      await expect(Cache.loadMany(rows)).resolves.toMatchSnapshot();
    });

    it('successfully update with operator', async () => {
      await Batch.batchUpdate(rows, { name: 'tiramisu' }, 99);
      await expect(Batch.loadMany(rows)).resolves.toMatchSnapshot();
    });

    it('operator is required', async () => {
      await expect(Batch.batchUpdate(rows, { name: 'sweets' }))
        .rejects.toEqual(new Error('operator is required'));
    });
  });

  describe('destroy', () => {
    it('successfully destory', async () => {
      const rows = ['1', '3', '5'];
      await Cache.batchDestroy(rows);
      await expect(Cache.loadMany(rows))
        .resolves.toEqual([null, null, null]);
    });

    it('successfully destory with disable soft delete', async () => {
      const rows = ['2', '4', '6'];
      await Logger.batchDestroy(rows, 99);
      await expect(Logger.loadMany(rows))
        .resolves.toEqual([null, null, null]);
    });

    it('successfully destory with operator', async () => {
      const rows = ['7', '8', '9'];
      await Batch.batchDestroy(rows, 99);
      await expect(Batch.loadMany(rows))
        .resolves.toEqual([null, null, null]);
    });

    it('operator is required', async () => {
      await expect(Batch.batchDestroy(['16']))
        .rejects.toEqual(new Error('operator is required'));
    });
  });
});
