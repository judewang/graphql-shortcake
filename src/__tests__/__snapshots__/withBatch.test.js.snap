// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Batch update successfully update 1`] = `
Array [
  Cache {
    "id": "iN7SNGO2W5Lur62xxuaxN",
    "name": "cupcake",
  },
  Cache {
    "id": "iN7SNGO2W5Lur62xxuaxP",
    "name": "cupcake",
  },
  Cache {
    "id": "iN7SNGO2W5Lur62xxuaxR",
    "name": "cupcake",
  },
]
`;

exports[`Batch update successfully update with operator 1`] = `
Array [
  Batch {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "id": "iN7LVv4xt1nXZSS0N6Y0v",
    "name": "tiramisu",
    "updatedAt": "[object DateTime]",
    "updatedBy": "99",
  },
  Batch {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "id": "iN7LVv4xt1nXZSS0N6Y0x",
    "name": "tiramisu",
    "updatedAt": "[object DateTime]",
    "updatedBy": "99",
  },
  Batch {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "id": "iN7LVv4xt1nXZSS0N6Y0z",
    "name": "tiramisu",
    "updatedAt": "[object DateTime]",
    "updatedBy": "99",
  },
]
`;
